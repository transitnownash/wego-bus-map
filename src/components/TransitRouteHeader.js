import React, { useState } from 'react';
import PropTypes from 'prop-types';
import L from 'leaflet';
import { Tooltip, OverlayTrigger } from 'react-bootstrap';
import { Link } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faExclamationTriangle } from '@fortawesome/free-solid-svg-icons';
import AlertModal from './AlertModal';
import busIcon from '../resources/bus.svg';
import trainIcon from '../resources/train.svg';
import './TransitRouteHeader.scss';

function TransitRouteHeader({ route, alerts, showRouteType }) {
  const [alertModalShow, setAlertModalShow] = useState(false);

  if (typeof route !== 'object' || !route.route_gid) {
    return (<div className="transit-route-header">Invalid route!</div>);
  }

  const routeStyle = {
    backgroundColor: `#${route.route_color}`,
    color: `#${route.route_text_color}`,
  };

  let vehicleIcon = busIcon;
  if (route?.route_type === '2') {
    vehicleIcon = trainIcon;
  }

  const routeAlertIcon = (
    <span className="transit-route-header-alert-trigger" onClick={() => setAlertModalShow(true)}>
      <FontAwesomeIcon icon={faExclamationTriangle} fixedWidth={true}></FontAwesomeIcon>
    </span>
  );

  return (
    <div className="transit-route-header d-flex" style={routeStyle} title={route.route_desc}>
      {showRouteType && (
        <div>
          <img className="me-2" style={{ height: '1.5rem' }} src={vehicleIcon} alt="Icon" title={`Route Type: ${route?.route_type}` } />
        </div>
      )}
      <div className="flex-grow-1 align-bottom">
        <Link to={`/routes/${route.route_gid}`}><span className="badge text-bg-light">{route.route_short_name}</span> {route.route_long_name}</Link>
      </div>
      <div>
        {alerts.length > 0 && (
          <div className="ms-2">
            {L.Browser.mobile === false ? (
              <OverlayTrigger placement={'top'} overlay={<Tooltip>{alerts.length > 1 ? `${alerts.length} alerts` : '1 alert'}</Tooltip>}>
                {routeAlertIcon}
              </OverlayTrigger>
            ) : (
              <>{routeAlertIcon}</>
            )}
            <AlertModal alerts={alerts} show={alertModalShow} onHide={() => setAlertModalShow(false)} routes={[route]}></AlertModal>
          </div>
        )}
      </div>
    </div>
  );
}

TransitRouteHeader.propTypes = {
  route: PropTypes.object.isRequired,
  alerts: PropTypes.array,
  showRouteType: PropTypes.bool,
};

TransitRouteHeader.defaultProps = {
  alerts: [],
  showRouteType: false,
};

export default TransitRouteHeader;
