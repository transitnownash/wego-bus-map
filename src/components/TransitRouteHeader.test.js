/* globals test, expect */

import React from 'react';
import { render, screen } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import TransitRouteHeader from './TransitRouteHeader';

const routeFixture = require('../fixtures/routes-4.json');

test('renders TransitRouteHeader', () => {
  const { container } = render(
    <Router>
      <TransitRouteHeader route={routeFixture} />
    </Router>,
  );
  expect(screen.getByText('SHELBY')).toBeInTheDocument();
  expect(container.querySelector('.transit-route-header')).toHaveStyle({ backgroundColor: 'rgb(117, 60, 190)', color: 'rgb(255, 255, 255)' });
  expect(container).toMatchSnapshot();
});

test('renders TransitRouteHeader', () => {
  const { container } = render(
    <Router>
      <TransitRouteHeader route={routeFixture} />
    </Router>,
  );
  expect(screen.getByText('SHELBY')).toBeInTheDocument();
  expect(container.querySelector('.transit-route-header')).toHaveStyle({ backgroundColor: 'rgb(117, 60, 190)', color: 'rgb(255, 255, 255)' });
  expect(screen.queryByAltText('Icon')).toBeNull();
  expect(container).toMatchSnapshot();
});

test('renders TransitRouteHeader with icon', () => {
  const { container } = render(
    <Router>
      <TransitRouteHeader route={routeFixture} showRouteType={true} />
    </Router>,
  );
  expect(screen.getByText('SHELBY')).toBeInTheDocument();
  expect(container.querySelector('.transit-route-header')).toHaveStyle({ backgroundColor: 'rgb(117, 60, 190)', color: 'rgb(255, 255, 255)' });
  expect(screen.getByAltText('Icon')).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

test('renders TransitRouteHeader with invalid route.', () => {
  const route = {};
  const { container } = render(
    <Router>
      <TransitRouteHeader route={route} />
    </Router>,
  );
  expect(screen.getByText('Not Assigned to Route')).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});
