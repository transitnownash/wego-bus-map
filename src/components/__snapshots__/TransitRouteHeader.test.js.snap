// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders TransitRouteHeader 1`] = `
<div>
  <div
    class="transit-route-header d-flex"
    style="background-color: rgb(117, 60, 190); color: rgb(255, 255, 255);"
  >
    <div
      class="flex-grow-1 align-bottom"
    >
      <a
        href="/routes/4"
      >
        <span
          class="badge text-bg-light"
        >
          4
        </span>
         
        SHELBY
      </a>
    </div>
    <div />
  </div>
</div>
`;

exports[`renders TransitRouteHeader 2`] = `
<div>
  <div
    class="transit-route-header d-flex"
    style="background-color: rgb(117, 60, 190); color: rgb(255, 255, 255);"
  >
    <div
      class="flex-grow-1 align-bottom"
    >
      <a
        href="/routes/4"
      >
        <span
          class="badge text-bg-light"
        >
          4
        </span>
         
        SHELBY
      </a>
    </div>
    <div />
  </div>
</div>
`;

exports[`renders TransitRouteHeader with icon 1`] = `
<div>
  <div
    class="transit-route-header d-flex"
    style="background-color: rgb(117, 60, 190); color: rgb(255, 255, 255);"
  >
    <div>
      <img
        alt="Icon"
        class="me-2"
        src="bus.svg"
        style="height: 1.5rem;"
        title="Route Type: 3"
      />
    </div>
    <div
      class="flex-grow-1 align-bottom"
    >
      <a
        href="/routes/4"
      >
        <span
          class="badge text-bg-light"
        >
          4
        </span>
         
        SHELBY
      </a>
    </div>
    <div />
  </div>
</div>
`;

exports[`renders TransitRouteHeader with invalid route. 1`] = `
<div>
  <div
    class="transit-route-header"
  >
    Not Assigned to Route
  </div>
</div>
`;
